source/client-side-encryption/client-side-encryption.md

```diff
@@ -0,0 +1,2483 @@
+
+A KMS provider (AWS KMS, Azure Key Vault, GCP KMS, the local provider, or KMIP) is used to decrypt data keys after
+fetching from the MongoDB Key Vault, and encrypt newly created data keys. Refer to [KMSProviders](#kmsproviders) for the
+shape of the KMS provider options.
+
+### mongocryptd
+
+`mongocryptd` is a singleton local process needed for auto-encryption when no [crypt_shared](#crypt_shared) library is
+used. It speaks the MongoDB wire protocol and the driver uses [mongocryptd](#mongocryptd) by connecting with a
+MongoClient. By default, if [crypt_shared](#crypt_shared) is unavailable, the driver should attempt to automatically
+spawn [mongocryptd](#mongocryptd). If the MongoClient is configured with `extraOptions.mongocryptdBypassSpawn` set to
+`true`, OR `bypassAutoEncryption` is set to `true`, OR `bypassQueryAnalysis` is set to `true` then the driver will not
+attempt to spawn [mongocryptd](#mongocryptd).
+
+The [mongocryptd](#mongocryptd) process is responsible for self terminating after idling for a time period. If
+[extraOptions.cryptSharedLibRequired](#extraoptions.cryptsharedlibrequired) is set to `true`, the driver will not
+connect to [mongocryptd](#mongocryptd) and instead rely on [crypt_shared](#crypt_shared) being available.
+
+### crypt_shared
+
+[crypt_shared](#crypt_shared) is a dynamically-loaded C++ library providing query analysis for auto-encryption. It
+replaces [mongocryptd](#mongocryptd) for performing query analysis to -
+Ref:`mark-up sensitive fields within a command <subtype6.intent-to-encrypt>`.
+
+Drivers are not required to load and interact with [crypt_shared](#crypt_shared) directly. Instead, they inform
+[libmongocrypt](#libmongocrypt) where to find [crypt_shared](#crypt_shared) and [libmongocrypt](#libmongocrypt) will
+handle [crypt_shared](#crypt_shared) communication automatically.
+
+See also: [Enabling crypt_shared](#enabling-crypt_shared) for information on using enabling the
+[crypt_shared](#crypt_shared) library.
+
+### libmongocrypt
+
+libmongocrypt is a C library providing crypto and coordination with external components.
+[Located here](https://github.com/mongodb/libmongocrypt).
+
+**libmongocrypt is responsible for…**
+
+- orchestrating an internal state machine.
+- asking the driver to perform I/O, then handling the responses.
+  - includes constructing KMS HTTP requests and parsing KMS responses.
+- doing encryption and decryption.
+- caching data keys.
+- caching results of listCollections.
+- creating key material.
+
+**The driver is responsible for…**
+
+- performing all I/O needed at every state:
+  - speaking to [mongocryptd](#mongocryptd) to mark commands (unless [crypt_shared](#crypt_shared) is used).
+  - fetching encrypted data keys from key vault collection (mongod).
+  - running listCollections on mongod.
+  - decrypting encrypted data keys with KMS over TLS.
+- doing I/O asynchronously as needed.
+
+See [Why require including a C library?](#why-require-including-a-c-library).
```

Comment:
```suggestion
See [Why require including a C library?](#why-require-including-a-c-library)
```

Trailing period doesn't add anything here and I noted you dropped it from similar lines.